core-js/src/definitions.ts

```diff
@@ -0,0 +1,1421 @@
+    const toReturn = [... this._referencers].map(r => {
+      SchemaElement.prototype['removeTypeReferenceInternal'].call(r, this);
+      return r;
+    });
+    this._referencers.clear();
+    return toReturn;
+  }
+
+  protected abstract removeInnerElements(): void;
+
+  toString(): string {
+    return this.name;
+  }
+}
+
+abstract class BaseNamedElementWithType<T extends Type, P extends SchemaElement<any, any> | Schema, Referencer> extends BaseNamedElement<P, Referencer> {
+  private _type?: T;
+
+  get type(): T | undefined {
+    return this._type;
+  }
+
+  set type(type: T | undefined) {
+    if (type) {
+      this.checkUpdate(type);
+    } else {
+      this.checkRemoval();
+    }
+    if (this._type) {
+      removeReferenceToType(this, this._type);
+    }
+    this._type = type;
+    if (type) {
+      addReferenceToType(this, type);
+    }
+  }
+
+  protected removeTypeReference(type: NamedType) {
+    if (this._type == type) {
+      this._type = undefined;
+    }
+  }
+}
+
+function buildError(message: string): Error {
+  // Maybe not the right error for this?
+  return new GraphQLError(message);
+}
+
+export class BuiltIns {
+  private readonly defaultGraphQLBuiltInTypes: readonly string[] = [ 'Int', 'Float', 'String', 'Boolean', 'ID' ];
+
+  addBuiltInTypes(schema: Schema) {
+    this.defaultGraphQLBuiltInTypes.forEach(t => this.addBuiltInScalar(schema, t));
+  }
+
+  addBuiltInDirectives(schema: Schema) {
+    for (const name of ['include', 'skip']) {
+      this.addBuiltInDirective(schema, name)
+        .addLocations('FIELD', 'FRAGMENT_SPREAD', 'FRAGMENT_DEFINITION')
+        .addArgument('if', new NonNullType(schema.booleanType()));
+    }
+    this.addBuiltInDirective(schema, 'deprecated')
+      .addLocations('FIELD_DEFINITION', 'ENUM_VALUE')
+      .addArgument('reason', schema.stringType(), 'No Longer Supported');
+    this.addBuiltInDirective(schema, 'specifiedBy')
+      .addLocations('SCALAR')
+      .addArgument('url', new NonNullType(schema.stringType()));
+  }
+
+  protected addBuiltInScalar(schema: Schema, name: string): ScalarType {
+    return schema.addType(new ScalarType(name, true));
+  }
+
+  protected addBuiltInObject(schema: Schema, name: string): ObjectType {
+    return schema.addType(new ObjectType(name, true));
+  }
+
+  protected addBuiltInUnion(schema: Schema, name: string): UnionType {
+    return schema.addType(new UnionType(name, true));
+  }
+
+  protected addBuiltInDirective(schema: Schema, name: string): DirectiveDefinition {
+    return schema.addDirectiveDefinition(new DirectiveDefinition(name, true));
+  }
+}
+
+export class Schema {
+  private _schemaDefinition: SchemaDefinition;
+  private readonly _builtInTypes: Map<string, NamedType> = new Map();
+  private readonly _types: Map<string, NamedType> = new Map();
+  private readonly _builtInDirectives: Map<string, DirectiveDefinition> = new Map();
+  private readonly _directives: Map<string, DirectiveDefinition> = new Map();
```

Comment:
Why the distinction between `builtInTypes`/`types` and `builtInDirectives`/`directives`? As built-in elements live in the same namespace as user-defined elements, can't they share a `Map`?

Reply:
This is somewhat opinionated but I have a strong hunch that much more code that will end up using this library will primarily not want to iterate about builtins (it may care they exists, but will not iterate/manipulate them). For instance, printing doesn't print those, composition will not merge those, any inaccessible removal won't care about those, etc... I'm sure at some point some code will want to iterate over all types/directives include builtins, but those example don't come easily to mind to me.

That's why `allChildElements` don't return builtins for instance (though I suppose maybe an optional argument to include them could make sense) and why `Schema.types` and `Schema.directives` don't include them as well.

So that's the justification for the external API, and the argument can be shorten as "I think it's more useful this way and that if we only always return all types/directives, everyone will have to filter built-ins out all the time".

Of course, internally, we can do whatever we want. We could have all in the same map, yet have the same API and filter built-in out within the accessor methods. But that just feel less efficient for no good reason.